api/src/lib/webhonc/index.ts

```diff
@@ -0,0 +1,161 @@
+import path from "node:path";
+import type { LibSQLDatabase } from "drizzle-orm/libsql";
+import WebSocket from "ws";
+import type { z } from "zod";
+import * as schema from "../../db/schema.js";
+import logger from "../../logger.js";
+import { resolveServiceArg } from "../../probe-routes.js";
+import {
+  executeProxyRequest,
+  handleFailedRequest,
+  handleSuccessfulRequest,
+} from "../proxy-request/index.js";
+import { WsMessageSchema } from "../types.js";
+import { resolveUrlQueryParams } from "../utils.js";
+import { getWebHoncConnectionId, setWebHoncConnectionId } from "./store.js";
+
+export function connectToWebhonc(
+  host: string,
+  db: LibSQLDatabase<typeof schema>,
+  wsConnections: Set<WebSocket>,
+) {
+  const protocol = host.startsWith("localhost") ? "ws" : "wss";
+  const socket = new WebSocket(`${protocol}://${host}/ws`);
+
+  socket.onopen = () => {
+    logger.debug(`Connected to the webhonc service at ${host}`);
+  };
+
+  socket.onmessage = async (event) => {
+    logger.debug("Received message from the webhonc service:", event.data);
+    await handleMessage(event, wsConnections, db);
+  };
+
+  socket.onclose = (event) => {
+    logger.debug(
+      "Webhonc connection closed",
+      event.reason,
+      event.code,
+      event.wasClean,
+    );
+  };
+
+  socket.onerror = (event) => {
+    logger.error("Webhonc connection error", event.message);
+  };
+
+  return socket;
+}
+
+async function handleMessage(
+  event: WebSocket.MessageEvent,
+  wsConnections: Set<WebSocket>,
+  db: LibSQLDatabase<typeof schema>,
+) {
+  const parsedMessage = WsMessageSchema.parse(
+    // it probably doesn't make sense that we're parsing the entire message and then re-serializing it
+    // for the request maker but this is the easiest way to move forward for now
+    JSON.parse(event.data.toString()),
+  );
+
+  const handler = messageHandlers[parsedMessage.event] as (
+    message: typeof parsedMessage,
+    wsConnections: Set<WebSocket>,
+    db: LibSQLDatabase<typeof schema>,
+  ) => Promise<void>;
+
+  if (handler) {
+    await handler(parsedMessage, wsConnections, db);
+  } else {
+    logger.error(`Unhandled event type: ${parsedMessage.event}`);
+  }
+}
+
+const messageHandlers: {
+  [K in z.infer<typeof WsMessageSchema>["event"]]: (
+    message: Extract<z.infer<typeof WsMessageSchema>, { event: K }>,
+    wsConnections: Set<WebSocket>,
+    db: LibSQLDatabase<typeof schema>,
+  ) => Promise<void>;
+} = {
+  trace_created: async (_message, _wsConnections, _db) => {
+    logger.info("trace_created message received, no action required");
```

Comment:
```suggestion
    logger.debug("trace_created message received, no action required");
```

probably don't want this to show up in the end user's terminal